beacon-chain/state/state-native/getters_participation.go

```diff
@@ -41,14 +44,22 @@ func (b *BeaconState) PreviousEpochParticipation() ([]byte, error) {
 // UnrealizedCheckpointBalances returns the total balances: active, target attested in
 // current epoch and target attested in previous epoch. This function is used to
 // compute the "unrealized justification" that a synced Beacon Block will have.
-func (b *BeaconState) UnrealizedCheckpointBalances() (uint64, uint64, uint64, error) {
-	if b.version == version.Phase0 {
-		return 0, 0, 0, errNotSupported("UnrealizedCheckpointBalances", b.version)
-	}
-
+func (b *BeaconState) UnrealizedCheckpointBalances(ctx context.Context) (uint64, uint64, uint64, error) {
 	b.lock.RLock()
 	defer b.lock.RUnlock()
 
+	if b.version == version.Phase0 {
+		vp, bp, err := precompute.New(ctx, b)
```

Comment:
This is a bit ugly, I am surprised you aren't running into cyclical dependency issues importing precompute here. Also I am pretty sure this is going to incur a deadlock here. Calling nested methods which are require a lock, is going to lead to deadlocks.